src/ir_passes/recomb.ml

```diff
@@ -0,0 +1,139 @@
+(* Turns rebuilt variants into casts of the same thing.
+   Also turns catchall cases into wildcard ones when the scrutinee is already a variable. *)
+
+open Ir_def
+open Ir
+open Source
+
+(* The transformation environment is trivial for now *)
+let empty_env () = ()
+
+(* The AST traversal *)
+
+(* Does a very simple thing (for now):
+ - spot `switch var {`
+ - where a variant match happens, that immediately recombines:
+    `case (#err err) { #err err }`
+ - replace with `case (#err _) { (prim cast) var }`
+ *)
+
+let rec t_exps env = List.map (t_exp env)
+
+and case env scrutinee case = match scrutinee, case with
+  (* switch v { ...; case (#err e) { #err e } } --> switch v { ...; case (#err _) { (prim cast) v } } *)
+  | { it = VarE _; _ }, { it = { pat = { it = TagP (ptag, ({it = VarP pid; _} as pv)); _ } as pat
+                               ; exp = { it = PrimE (TagPrim etag, [{ it = VarE eid; _}]); _ } as exp }; _ }
+    when ptag = etag && pid = eid ->
+    { case with it = { pat = { pat with it = TagP (ptag, {pv with it = WildP})}
+                     ; exp = { exp with it = cast scrutinee.note.Note.typ exp.note.Note.typ
+                                             {(t_exp env scrutinee) with at = exp.at} } } }
+  (* switch v { ...; case p p } --> switch v { ...; case _ v } *)
```

Comment:
This isn't safe unless it's the last case, but I don't see where you require that.